Algorithm for Finding Valid Cycles in an Improvement Graph

christophgruene opened the pull request:
This pull request fixes #588.
There is one question left. I introduced a class LabeledPath, in which I maintain necessary information for the algorithm. In these paths I do not have to store the edges, hence, I did not implement the interface GraphPath. Should I do this from your point of view or is it ok to leave it likes it is?

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,161 @@
+/**
+ * Implementation of an algorithm for the local augmentation problem for the cyclic exchange neighborhood,
+ * i.e. it finds subset-disjoint negative cycles, based on the
+ * <a href ="http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.15.6758&rep=rep1&type=pdf">paper</a>
```
Please follow the citation conventions in

https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation

This is preferred over a link, since links may break at any time.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,161 @@
+    public LabeledPath<V> getLocalAugmentationCycle() {
+
+        int k = 1;
+        LabeledPath<V> C = new LabeledPath<>();
```
Variable names should always start with a lowercase letter; only class names start with an uppercase letter.  (This applies to some of your other pull requests as well.)

christophgruene replied:
This is true and I know the convention but I want to use the notation of the paper. However, I will change this here.

christophgruene commented:
In the future, I want to implement a large neighborhood search for the capacitated minimum spanning tree (https://en.wikipedia.org/wiki/Capacitated_minimum_spanning_tree), in which I use this algorithm. I think, this actually fits the scope of this library. Then, this algorithm is used by an actual graph problem.

Nevertheless, this algorithm is interesting for many partitioning problems, which use an improvement graph in order to find a solution.

christophgruene commented:
@jsichi I want to implement the CMST optimization algorithm, which uses the valid cycle detection algorithm, based on the same paper. Should I include it in this PR or should I open a new one (then, I will open a new issue addressing CMST as well)?

jsichi commented:
Since you're planning to follow up with the others, we can finish the review for this one and merge it first.  I'll do another review pass.

Meanwhile, @christophgruene could you let me know your email address?  You can email me at jsichi at gmail.com.  To help give us context during code reviews, I'd like to get some more information about your team and your overall project plan, but I don't have a working email address for anyone.  (I tried emailing @pdelvo a few days ago but didn't see a response.)  Thanks!

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,335 @@
+            while(!pathsLengthK.isEmpty()) {
+
+                // go through all valid paths of length k
+                for(Iterator<LabeledPath<V>> it = pathsLengthK.iterator(); it.hasNext();) {
```
Why do you need to remove the paths iteratively?  You're just going to discard the entire pathsLengthK at the end anyway.  Skip that and then use a compact for loop `for (LabeledPath<V> path : pathLengthK)`.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,335 @@
+     * @param lengthBound the upper bound for the length of cycles to detect
+     * @param labels the labels of the vertices encoding the subsets of vertices
+     */
+    public AhujaOrlinSharmaCyclicExchangeLocalAugmentation(Graph<V, E> graph, int lengthBound, Map<V, Integer> labels) {
```
Can you think of any sanity-checking you should be doing on the inputs here?

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,335 @@
+     * @param path the currently calculated path
+     * @param PathsLengthKplus1 all before calculated paths of length k + 1
+     */
+    private void testDomination(LabeledPath<V> path, Set<LabeledPath<V>> PathsLengthKplus1) {
```
variable naming convention

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,335 @@
+                break;
+            }
+            if(path.dominates(pathInSetKplus1)) {
+                PathsLengthKplus1.remove(pathInSetKplus1);
```
Why is this not at risk of ConcurrentModificationException as well?  Are you somehow relying on the LinkedHashSet ordering?  That's not a good idea.

This is a case where it.remove can be used instead (since pathInSetKplus1 was the last return value of it.next).

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,335 @@
+                             * check if paths are dominated, i.e. if the path is definitely worse than other paths
+                             * and does not have to be considered in the future
+                             */
+                            testDomination(path, pathsLengthKplus1);
```
Shouldn't you be testing newPath rather than path?  If that's true, then it's probably simpler to defer adding it until after testDomination.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,335 @@
+    private void testDomination(LabeledPath<V> path, Set<LabeledPath<V>> PathsLengthKplus1) {
+        boolean removePath = false;
+
+        for(Iterator<LabeledPath<V>> it = PathsLengthKplus1.iterator(); it.hasNext();) {
```
I'm pretty sure you could use an index structure of some kind to avoid having to iterate over the entire set here.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,335 @@
+    public static class LabeledPath<V> implements Cloneable {
+
+        private LinkedList<V> vertices;
+        private Map<V, Integer> labels;
```
I believe this path-level map is always a submap of the graph-level labels map (i.e. we never override labels when constructing a path).  If that's true, wouldn't it be simpler to just access the graph-level map (which you can do by making LabeledPath non-static)?

christophgruene commented:
Sorry for this mass of needed changes, I fixed them all. I changed `Set<LabeledPath<V>>` to `Map<Pair<V, V>, TreeSet<LabeledPath<V>>>`, whereby the tree set is sorted by the cost of the paths. The checks in the domination tests decrease because the domination relation exists only on paths that have the same head and tail. I do not know if it is important to use a tree set instead of a set in a practical setting. However, this reduces the domination test checks to optimal in every tree set but this has to be paid by higher insertion cost.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,343 @@
+
+                    // the path builds a valid negative cycle
+                    if (graph.containsEdge(tail, head)
+                            && path.getCost() + graph.getEdgeWeight(graph.getEdge(tail, head)) < 0
```
Don't duplicate the computation of the new cost.  Also, I don't understand why both checks are needed.  We return as soon as we hit a negative cycle, so if the newly calculated cost is negative, then it is definitely less than the previously computed cycleResult cost.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,343 @@
+            }
+        }
+
+        while (k < lengthBound && cycleResult.getCost() >= 0) {
```
Won't cycleResult always be nonnegative here since if we hit a negative cycle inside of this loop, we return immediately?  In fact, I can't understand why cycleResult needs to be maintained across loops at all.  It should only be constructed directly at the point of return unless it's being used for some kind of minimization/maximization.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,343 @@
+
+                    for (E e : graph.outgoingEdgesOf(tail)) {
+                        V currentVertex = graph.getEdgeTarget(e);
+                        if (!path.getLabels().contains(labels.get(currentVertex)) && path.getCost() + graph.getEdgeWeight(e) < 0) { // extend the path if the extension is still negative a correctly labeled
```
"and correctly labeled"

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,343 @@
+        Map<Pair<V, V>, TreeSet<LabeledPath<V>>> pathsLengthKplus1 = new LinkedHashMap<>();
+
+        // initialize pathsLengthK for k = 1
+        for(E e : graph.edgeSet()) {
```
Is the input graph allowed to have self-loops?  If so, then what happens if one of these edges already forms a cycle?  If not, then that should be a sanity check.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,343 @@
+                        V currentVertex = graph.getEdgeTarget(e);
+                        if (!path.getLabels().contains(labels.get(currentVertex)) && path.getCost() + graph.getEdgeWeight(e) < 0) { // extend the path if the extension is still negative a correctly labeled
+                            LabeledPath<V> newPath = path.clone();
+                            newPath.addVertex(currentVertex, graph.getEdgeWeight(e), labels.get(currentVertex));
```
Don't repeat the label lookup

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,343 @@
+     * Constructs an algorithm with given inputs
+     *
+     * @param graph the (improvement) graph on which to calculate the local augmentation
+     * @param lengthBound the upper bound for the length of cycles to detect
```
Specify whether this is `<` or `<=`

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,343 @@
+ * A composite very large-scale neighborhood structure for the capacitated minimum spanning tree problem,
+ * Operations Research Letters, Volume 31, Issue 3, 2003, Pages 185-194, ISSN 0167-6377,
+ * https://doi.org/10.1016/S0167-6377(02)00236-5. (http://www.sciencedirect.com/science/article/pii/S0167637702002365)
+ *
```
Add a precise definition of subset-disjoint here, so that a user can understand the algorithm input/output specification without having to read the paper.  Also specify *which* such cycle it finds; i.e. is it the shortest by edge length?  least negative?  arbitrary?  etc.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,343 @@
+                V targetVertex = graph.getEdgeTarget(e);
+                pathVertices.add(sourceVertex);
+                pathVertices.add(targetVertex);
+                pathLabels.add(labels.get(sourceVertex));
```
Why is it that these initial paths are allowed to have the same label for source and target, but in later rounds, we require each newly added vertex to have a label distinct from those already in the path?

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,343 @@
+     * @param labels the labels of the vertices encoding the subsets of vertices
+     */
+    public AhujaOrlinSharmaCyclicExchangeLocalAugmentation(Graph<V, E> graph, int lengthBound, Map<V, Integer> labels) {
+        this.graph = Objects.requireNonNull(graph, "Graph cannot be null");
```
Are there any other restrictions on the input graph, e.g. directed vs undirected?  Are multiple edges between vertex pairs allowed?  Doesn't the graph need to be weighted?

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,343 @@
+            try {
+                LabeledPath<V> newLabeledPath = TypeUtil.uncheckedCast(super.clone());
+                newLabeledPath.vertices = (ArrayList<V>) this.vertices.clone();
+                newLabeledPath.labels = new HashSet<>(this.labels);
```
Why not clone the labels too?

christophgruene commented:
I have to correct the deletion of the new created path p_k+1 of length k+1. You are right that we can add the set to the key since the length of both label sets has length k+1. But it has to be checked whether p_k+1 is dominated by a path from the set P_k of paths of length k, too.

christophgruene commented:
I added a new class PathSetKey to have a triple for the three objects. Furthermore, I introduced a method that checks if one path of length k (corresponding to head, tail and labels) exists that dominates the new created path of length k+1. The problem with cycles of length 1 is corrected as well.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,427 @@
+     * @author Christoph Grüne
+     * @since June 7, 2018
+     */
+    private static class PathSetKey<V> {
```
Don't you have to override equals/hashCode?  If the answer is "yes", then you might want to think about whether your unit tests (which are currently passing) are providing sufficient coverage.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,427 @@
+        TreeSet<LabeledPath<V>> pathsToCheck = pathsLengthKplus1.get(PathSetKey.of(path.getHead(), path.getTail(), path.getLabels()));
+        if(pathsToCheck != null) {
+            // simulates LabeledPath.dominates(LabeledPath) by using the index structure
+            for (Iterator<LabeledPath<V>> it = pathsToCheck.descendingIterator(); it.hasNext(); ) {
```
Now that we have the comparison purely on cost, can you think of any more efficient way to do this using the TreeSet interface instead of iteration?  (I don't know the answer since I'm not sure whether it's possible to efficiently remove the entire tail of the tree.)

christophgruene replied:
The question is whether `Subset.clear()` runs in O(1). I cannot find anything to this question. Then, it is possible to call `tailset(E)` and call `clear()` on it.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,427 @@
+            TreeSet<LabeledPath<V>> pathsToCheck = pathsLengthK.get(PathSetKey.of(path.getHead(), path.getTail(), modifiableLabelSet));
+            if (pathsToCheck != null) {
+                // simulates LabeledPath.dominates(LabeledPath) by using the index structure
+                for (Iterator<LabeledPath<V>> it = pathsToCheck.descendingIterator(); it.hasNext(); ) {
```
This one can definitely be optimized using TreeSet.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,427 @@
+         * @param path2 the second labeled path
+         * @return whether this instance dominates <code>path2</code>
+         */
+        protected boolean dominates(LabeledPath<V> path2) {
```
This is no longer called at all except in comments.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,445 @@
+        }
+
+        @Override
+        public int hashCode() {
```
It's cleaner to use Objects.hash

https://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#hash(java.lang.Object...)

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,445 @@
+                     */
+                    break;
+                }
+                if (path.getCost() < pathInSetKplus1.getCost()) {
```
I concur that calling clear() on a submap view is probably no better than iteration (I think under the covers that is how it ends up implemented).

There is one other way to do it though, which is to replace the original TreeSet with the subset to retain.  The implementation for this is efficient because it just keeps begin/end pointers into the original set (and doing this more than once just moves the pointers).

But stepping back for a moment, is there even a need to keep a TreeSet at all?  In the end, don't we only keep the path of lowest cost?

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,419 @@
+/*
+ * (C) Copyright 2003-2018, by Christoph Grüne and Contributors.
```
Copyright date should start from 2018.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,419 @@
+    }
+
+    /**
+     * Implementation of an key for the path maps.
```
"a key"

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,419 @@
+        int k = 1;
+
+        /*
+         * Store the path sets via a with Pair<V, V> as for the head and tail of the path such that the domination tests runs faster.
```
Pair and TreeSet are no longer used.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,419 @@
+     *
+     * @return whether <code>path</code> dominates the current minimal cost path with the same head, tail and label set.
+     */
+    private boolean removeDominatedPathsOfLengthKplus1(LabeledPath<V> path, Map<PathSetKey<V>, LabeledPath<V>> pathsLengthKplus1) {
```
Since this method no longer does the actual removal, it should be renamed from "remove" to "check".

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,419 @@
+     *
+     * @return whether <code>path</code> is dominated by some path in <code>pathsLengthK</code>
+     */
+    private boolean removeDominatedPathsOfLengthK(LabeledPath<V> path, Map<PathSetKey<V>, LabeledPath<V>> pathsLengthK) {
```
Same here.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,419 @@
+    }
+
+    /**
+     * Checks whether <code>path</code> is dominated by some path in the before calculated set of paths of length k.
```
"before" -> "previously"

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,419 @@
+     * be considered in further calculations.
+     *
+     * @param path the currently calculated path
+     * @param pathsLengthK all before calculated paths of length k
```
"before" -> "previously"

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,419 @@
+    }
+
+    /**
+     * Adds a path to the data structure <code>paths</code>, which contains all paths indexed by their head and tail, such that they are sorted by their cost.
```
I suggest renaming this method to "updatePathIndex" since now it does an add-or-replace rather than just an add.

jsichi replied:
This comment needs to be updated.

christophgruene commented:
@jsichi, thank you very much for the great support. I promise, the next PRs will be more sophisticated!

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,420 @@
+     * @author Christoph Grüne
+     * @since June 7, 2018
+     */
+    private static class LabeledPath<V> implements Cloneable {
```
Move this inner class at the very bottom of your class. When someone reads the code, they should get a quick feeling of what is going on, by simply looking at the beginning of the class.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,420 @@
+         * @param cost the cost of the edges connecting the vertices
+         * @param labels the mapping of the vertices to labels (subsets)
+         */
+        protected LabeledPath(ArrayList<V> vertices, double cost, HashSet<Integer> labels) {
```
Why is this protected? Do you use this class as a Base class somewhere? If not, then make it public.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,420 @@
+         * @param edgeCost the cost of the edge connecting the last vertex of the path and the new vertex
+         * @param label the label of the new vertex
+         */
+        protected void addVertex(V v, double edgeCost, int label) {
```
Same here, and in all other such methods.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,420 @@
+     * @author Christoph Grüne
+     * @since June 7, 2018
+     */
+    private static class PathSetKey<V> {
```
Move this class also to the end.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,420 @@
+         * @param <V> the vertex type
+         * @return a new PathSetKey object
+         */
+        protected static <V> PathSetKey<V> of(V head, V tail, Set<Integer> labels) {
```
Make this public

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,420 @@
+    /**
+     * the input graph
+     */
+    private Graph<V, E> graph;
```
These fields should be at the beginning of the class, follows by the public constructors.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java:
```diff
@@ -0,0 +1,420 @@
+import java.util.*;
+
+/**
+ * Implementation of an algorithm for the local augmentation problem for the cyclic exchange neighborhood,
```
The main problem with this implementation is that it is useable only by people who have read the relevant papers about the CMST problem and have also understood the concept of the Improvement Graph and the concept of “subset-disjoint" cycles in the improvement graph.

All these need to be clearly explained inside the Javadoc. 
 - What is the improvement graph? 
 - What is the "subset-disjoint" cycle? 
 - What are the labels that you providing to the algorithm? 
 - What kind of algorithm has produced these labels?
 - Are then any invariants that must be valid for the input?

The big picture is missing here, which means that it is highly unlikely that someone else will use this implementation.

christophgruene commented:
I can provide a more understandable description for the usages of the algorithm. The point is that this algorithm cannot only be used for CMST but also for all partition based neighborhood heuristics, of course as a subroutine.
In order to understand the usage of the algorithm, one has to understand what an improvement graph is (and I think only those people will use the algorithm). I can provide a description for improvement graphs, too.
Should I write all of these things in the head javaDoc of the class?

jsichi commented:
Based on the comments from @d-michail let's do this.  Instead of me merging this one, you can open a new pull request which combines this one with your CMST implementation.  Then we'll review that one and close this one.  (I'd prefer not to keep reviewing in this PR since the history is already getting long.)

That way you can document everything just once.  Put common definitions such as improvement graph into only one of the classes, and then use `@see` or `@link` Javadoc tags to refer to it from others.

christophgruene commented:
@d-michail What do you mean with 
> (as proposed in #634 but without the heuristics)?

The Esau-Williams algorithm is a greedy heuristic and the neighborhood search of Ahuja et al. is a local search heuristic, which uses the Esau-Williams algorithms. Both do not provide an approximation ratio, but Ahuja's local search is good in a pratical setting. Do you not want to have these heuristics (for the CMST) in this library?

d-michail commented:
I meant the additional heuristics. Your contribution is very high quality and we definitely would like to have what you are proposing in the library. 

However, I would prefer to do it step-wise, and with proper documentation and context, so that others can also benefit from these implementations.

I would suggest to start by introducing a nice interface for the CMST problem (inside the `org.jgrapht.alg.interfaces` package). Then by submitting the main algorithm which builds the improvement graph and calls the heuristics you are developing.  Note that the end-user of your implementation should have no detailed knowledge of the improvement graph or any details of the algorithms.

jsichi commented:
@christophgruene I htink you can close this one now since it's superseded by #655
